tests/src/opaque_test_driver.c

```diff
@@ -0,0 +1,338 @@
+        ( psa_get_key_type( attributes ) !=
+          PSA_KEY_TYPE_ECC_KEY_PAIR( PSA_ECC_CURVE_SECP_R1 ) ) )
+        return( PSA_ERROR_NOT_SUPPORTED );
+
+    if( ( psa_get_key_bits( attributes ) != 128 ) &&      // AES-128
+        ( psa_get_key_bits( attributes ) != 192 ) &&      // AES-192
+        ( psa_get_key_bits( attributes ) != 256 ) )       // AES-256
+        return( PSA_ERROR_NOT_SUPPORTED );
+
+    if( psa_get_key_bits( attributes ) != PSA_BYTES_TO_BITS( in_length ) )
+        return( PSA_ERROR_INVALID_ARGUMENT );
+
+    if( OPAQUE_TEST_DRIVER_KEYHEADER_SIZE + in_length > out_size )
+        return( PSA_ERROR_BUFFER_TOO_SMALL );
+
+    strcpy( (char *) out, OPAQUE_TEST_DRIVER_KEYHEADER );
+
+    /* Obscure key slightly. */
+    rot13( in, in_length, out + OPAQUE_TEST_DRIVER_KEYHEADER_SIZE );
+
+    *out_length = in_length + OPAQUE_TEST_DRIVER_KEYHEADER_SIZE;
+
+    return( PSA_SUCCESS );
+}
+
+psa_status_t opaque_test_driver_sign_hash(
+                                        const psa_key_attributes_t *attributes,
+                                        const uint8_t *key,
+                                        size_t key_length,
+                                        psa_algorithm_t alg,
+                                        const uint8_t *hash,
+                                        size_t hash_length,
+                                        uint8_t *signature,
+                                        size_t signature_size,
+                                        size_t *signature_length )
+{
+    #define OPQ_BUFSIZE 64
+    size_t key_buffer_length;
+    psa_key_handle_t handle = 0;
+    uint8_t key_buffer[OPQ_BUFSIZE];
+    psa_status_t status = PSA_SUCCESS;
+
+    OPQTD_VALIDATE_RET( attributes != NULL );
+    OPQTD_VALIDATE_RET( key != NULL );
+    OPQTD_VALIDATE_RET( hash != NULL );
+    OPQTD_VALIDATE_RET( signature != NULL );
+    OPQTD_VALIDATE_RET( signature_length != NULL );
+
+    if( key_length <= OPAQUE_TEST_DRIVER_KEYHEADER_SIZE )
+        return( PSA_ERROR_INVALID_ARGUMENT );
+
+    status = opaque_test_driver_export_public_key( attributes,
+                                                   key,
+                                                   key_length,
+                                                   key_buffer,
+                                                   OPQ_BUFSIZE,
+                                                   &key_buffer_length );
+    if( status != PSA_SUCCESS )
+        return( status );
+
+    status = psa_import_key( attributes,
+                             key_buffer, key_buffer_length,
+                             &handle );
+    if( status != PSA_SUCCESS )
+        return( status );
+
+    if( PSA_SIGN_OUTPUT_SIZE( psa_get_key_type( attributes ),
+                              psa_get_key_bits( attributes ),
+                              alg ) > signature_size )
+    {
+        psa_destroy_key( handle );
+        return( PSA_ERROR_BUFFER_TOO_SMALL );
```

Comment:
We generally prefer this style, because it leads to smaller code and less risk of forgetting some cleanup:

```suggestion
        status = PSA_ERROR_BUFFER_TOO_SMALL;
        goto exit;
```

and then call `psa_destroy_key` after `exit:`.